src/platform_impl/macos/app_state.rs

```diff
@@ -85,27 +85,35 @@ impl<T> Debug for EventLoopHandler<T> {
 }
 
 impl<T> EventHandler for EventLoopHandler<T> {
-    fn handle_nonuser_event(&mut self, event: Event<Never>, control_flow: &mut ControlFlow) {
+    fn handle_nonuser_event(&mut self, event: Event<Never>) {
         self.with_callback(|this, mut callback| {
-            if let ControlFlow::ExitWithCode(code) = *control_flow {
-                // XXX: why isn't event dispatching simply skipped after control_flow = ExitWithCode?
-                let dummy = &mut ControlFlow::ExitWithCode(code);
-                (callback)(event.userify(), &this.window_target, dummy);
+            let is_exit = if let ControlFlow::ExitWithCode(code) = HANDLER.control_flow() {
+                Some(code)
             } else {
-                (callback)(event.userify(), &this.window_target, control_flow);
+                None
+            };
+
+            (callback)(event.userify(), &this.window_target);
+
+            if let Some(code) = is_exit {
+                HANDLER.set_control_flow(ControlFlow::ExitWithCode(code))
             }
```

Comment:
Nit: I think it makes sense to move this check to `set_control_flow` instead, that way we're more sure it can never get reset after being set to `ExitWithCode`.

Reply:
Good idea!

Reply:
Done.